File: desktop-mend/src/test/java/commands/EncryptFromStdInTest.java
package commands;

import co.samco.mend4.core.Settings;
import co.samco.mend4.core.exception.CorruptSettingsException;
import co.samco.mend4.desktop.commands.Encrypt;
import co.samco.mend4.desktop.commands.EncryptFromStdIn;
import co.samco.mend4.desktop.exception.SettingRequiredException;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.mockito.ArgumentCaptor;

import javax.crypto.BadPaddingException;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.NoSuchPaddingException;
import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.security.InvalidAlgorithmParameterException;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.security.spec.InvalidKeySpecException;
import java.util.Arrays;
import java.util.Collections;

import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class EncryptFromStdInTest extends TestBase {
    private EncryptFromStdIn encrypt;

    @Before
    public void setup() {
        super.setup();
        try {
            when(settings.valueSet(Settings.Name.ENCDIR)).thenReturn(true);
            when(settings.valueSet(Settings.Name.LOGDIR)).thenReturn(true);
        } catch (IOException e) {
            e.printStackTrace();
        }
        encrypt = new EncryptFromStdIn(settings, settingsHelper, log, strings, cryptoHelper, inputHelper,
                osDao, fileResolveHelper);
    }

    @Test
    public void testCommandWithSettingsDependencies() throws IOException, SettingRequiredException {
        super.testCommandWithSettingsDependencies(encrypt);
    }

    @Test
    public void encryptFromStdIn() throws IOException, NoSuchAlgorithmException, InvalidKeyException,
            InvalidAlgorithmParameterException, NoSuchPaddingException, BadPaddingException, CorruptSettingsException,
            InvalidKeySpecException, IllegalBlockSizeException {
        encryptFromStdIn(false);
    }

    @Test
    public void encryptFromStdInAppend() throws IOException, NoSuchAlgorithmException, InvalidKeyException,
            InvalidAlgorithmParameterException, NoSuchPaddingException, BadPaddingException, CorruptSettingsException,
            InvalidKeySpecException, IllegalBlockSizeException {
        encryptFromStdIn(true);
    }

    public void encryptFromStdIn(boolean appendFlag) throws IOException, NoSuchPaddingException, NoSuchAlgorithmException,
            InvalidKeyException, InvalidAlgorithmParameterException, IllegalBlockSizeException, BadPaddingException,
            CorruptSettingsException, InvalidKeySpecException {
        String input = "hi\rhi again\nhi once more";
        InputStream inputStream = new ByteArrayInputStream(input.getBytes(StandardCharsets.UTF_8));
        when(osDao.getStdIn()).thenReturn(inputStream);
        encrypt.execute(appendFlag ? Arrays.asList(Encrypt.APPEND_FLAG) : Collections.emptyList());
        ArgumentCaptor<char[]> encryptOutput = ArgumentCaptor.forClass(char[].class);
        verify(osDao).getStdIn();
        verify(cryptoHelper).encryptTextToLog(encryptOutput.capture(), eq(appendFlag));
        String expectedText = "hi" + System.getProperty("line.separator") + "hi again" + System.getProperty("line.separator") + "hi once more";
        Assert.assertEquals(expectedText, new String(encryptOutput.getValue()));
    }

}

File: desktop-mend/src/test/java/commands/EncryptFromStdInTest.kt
package commands

import co.samco.mend4.desktop.commands.Encrypt
import co.samco.mend4.desktop.commands.EncryptFromStdIn
import co.samco.mend4.desktop.dao.SettingsDao
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.mockito.ArgumentCaptor
import com.nhaarman.mockitokotlin2.eq
import com.nhaarman.mockitokotlin2.verify
import com.nhaarman.mockitokotlin2.whenever
import java.io.ByteArrayInputStream
import java.io.IOException
import java.io.InputStream
import java.nio.charset.StandardCharsets

class EncryptFromStdInTest : TestBase() {
    private lateinit var encrypt: EncryptFromStdIn

    @Before
    override fun setup() {
        super.setup()
        try {
            whenever(settings.getValue(SettingsDao.ENC_DIR)).thenReturn("encdir")
            whenever(settings.getValue(SettingsDao.LOG_DIR)).thenReturn("decdir")
        } catch (e: IOException) {
            e.printStackTrace()
        }
        encrypt = EncryptFromStdIn(
            settings = settings,
            settingsHelper = settingsHelper,
            log = log,
            strings = strings,
            cryptoHelper = cryptoHelper,
            inputHelper = inputHelper,
            fileResolveHelper = fileResolveHelper
        )
    }

    @Test
    fun testCommandWithSettingsDependencies() {
        super.testCommandWithNoSettingsDependencies(encrypt)
    }

    @Test
    fun encryptFromStdIn() {
        encryptFromStdIn(false)
    }

    @Test
    fun encryptFromStdInAppend() {
        encryptFromStdIn(true)
    }

    private fun encryptFromStdIn(appendFlag: Boolean) {
        val input = "hi\rhi again\nhi once more"
        val inputStream: InputStream =
            ByteArrayInputStream(input.toByteArray(StandardCharsets.UTF_8))
        System.setIn(inputStream)
        encrypt.execute(if (appendFlag) listOf(Encrypt.APPEND_FLAG) else emptyList())
        val encryptOutput: ArgumentCaptor<CharArray> =
            ArgumentCaptor.forClass(CharArray::class.java)
        verify(cryptoHelper).encryptTextToLog(encryptOutput.capture(), eq(appendFlag))
        val expectedText =
            "hi" + System.getProperty("line.separator") + "hi again" + System.getProperty("line.separator") + "hi once more"
        Assert.assertEquals(expectedText, String(encryptOutput.value))
    }
}